include/libdnf/utils/exception.hpp

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright (C) 2020 Red Hat, Inc.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef LIBDNF_UTILS_EXCEPTION_HPP
+#define LIBDNF_UTILS_EXCEPTION_HPP
+
+
+#include <stdexcept>
+#include <system_error>
+
+namespace libdnf {
+
+/// Base class of libdnf exceptions. Each exception define at least domain_name, name, and description.
+class Exception : public std::runtime_error {
+public:
+    using runtime_error::runtime_error;
+    virtual const char * get_domain_name() const noexcept { return "libdnf"; }
+    virtual const char * get_name() const noexcept { return "Exception"; }
+    virtual const char * get_description() const noexcept { return ""; }
+};
+
+class LogicError : public Exception {
+public:
+    using Exception::Exception;
+    const char * get_name() const noexcept override { return "LogicError"; }
+    const char * get_description() const noexcept override { return "General LogicError exception"; }
+};
+
+class RuntimeError : public Exception {
```

Comment:
`RuntimeError` vs. `std::runtime_error`. They look very similar, but hope that they are not too similar.

Reply:
In fact they are similar in the meaning. I created two classes - `LogicError` and `RuntimeError`. The idea is `LoggicError` as base class for program (bugs in the source code) errors and `RuntimeError` as base class fore another bugs.
But I'm not sure if we will be abble to determine the right cause of the error. May be, in most cases we will use `RuntimeError` base class.